### PR TITLE
Rename recieve to Recv, add move semantics for SendTagged

### PR DESCRIPTION
**Pull request by ddn0**
I before E except after C.

It seems like "recv" is a common abbreviation for receive in newer code
so use that in place of the misspelt recieve.

Also,

- All code refers to DeSerializeBuffer and SerializeBuffer through their
  typedefs RecvBuffer and SendBuffer. Rename the classes themselves to
  avoid multiple names for the same thing
- Remove unused constructors for SendBuffer and RecvBuffer
- Remove unused timeLock in NetworkBuffer
- Remove the incantation "decl(net.recieveTagged(...))" in favor of a
  specific type, RecvResult. Though it is possible that NetworkInterface
  subclasses override RecvTagged with a covariant return value (i.e.,
  return a subclass of RecvResult), RecvResult itself is std::pair and
  it is not very likely that anyone would want to subclass std::pair
  directly.

### CODE REVIEW

File: libdist/include/galois/runtime/Network.h [hunk 1]
```diff
@@ -78,27 +77,21 @@ class GALOIS_EXPORT NetworkInterface : public CommBackend {
   virtual ~NetworkInterface();
 
   //! Send a message to a given (dest) host.  A message is simply a
-  //! landing pad (recv, funciton pointer) and some data (buf)
-  //! on the receiver, recv(buf) will be called durring handleReceives()
+  //! landing pad and some data (buf).
+  //! On the receiver, recv(buf) will be called during handleRecvs()
   //! buf is invalidated by this operation
   void sendMsg(
```
Comment (roshandathathri) [inline libdist/include/galois/runtime/Network.h:83]: I would prefer this function to be dropped. We haven't used it in a long time. This was designed for the _active message_ model of communication and I don't see a use for it in the immediate future.

File: libdist/include/galois/runtime/Network.h [hunk 2]
```diff
@@ -78,27 +77,21 @@ class GALOIS_EXPORT NetworkInterface : public CommBackend {
...
-  void sendSimple(
-      uint32_t dest, void (*recv)(uint32_t, Args...), Args... param);
-
   //! Send a message to a given (dest) host.  A message is simply a
   //! tag (tag) and some data (buf)
-  //! on the receiver, buf will be returned on a receiveTagged(tag)
+  //! on the receiver, buf will be returned on a recvTagged(tag)
   //! buf is invalidated by this operation
   virtual void sendTagged(
       uint32_t dest, uint32_t tag, SendBuffer& buf, int type = 0) = 0;
```
Comment (roshandathathri) [inline libdist/include/galois/runtime/Network.h:91]: `SendBuffer&& buf` is precise. We can defer that as that will involve changing a lot of code.

**Review by tylershunt — APPROVED** [review]
LGTM I always spell Receive wrong then fix it only to realize I had it right the first time.

File: libdist/include/galois/runtime/Network.h [hunk 3]
```diff
@@ -41,10 +41,9 @@
 
 namespace galois::runtime {
 
-//! typedef for buffer that stores data to be sent out
-using SendBuffer = SerializeBuffer;
-//! typedef for buffer that received data is saved into
-using RecvBuffer = DeSerializeBuffer;
+//! The result of receiving from a network interface is a pair of sending host
+//! and data.
+using RecvResult = std::optional<std::pair<uint32_t, RecvBuffer>>;
```
Comment (roshandathathri) [inline libdist/include/galois/runtime/Network.h:46]: Isn't this better as a type within the NetworkInterface class?

File: libdist/include/galois/runtime/Serialize.h [hunk 4]
```diff
@@ -52,57 +52,30 @@ namespace runtime {
...
  */
-class DeSerializeBuffer {
+class RecvBuffer {
   //! type of data buffer
-  // using vTy = std::vector<uint8_t>;
   using vTy = galois::PODResizeableArray<uint8_t>;
   //! the actual data stored in this buffer
   vTy bufdata;
-  int offset;
+  unsigned offset{0};
```
Comment (roshandathathri) [inline libdist/include/galois/runtime/Serialize.h:60]: Why not uint32_t?

File: libdist/include/galois/runtime/Serialize.h [hunk 5]
```diff
@@ -174,27 +148,19 @@ class DeSerializeBuffer {
...
+class SendBuffer {
   //! type of data buffer
-  // using vTy = std::vector<uint8_t>;
   using vTy = galois::PODResizeableArray<uint8_t>;
   //! the actual data stored in this buffer
   vTy bufdata;
 
 public:
-  //! default constructor
-  SerializeBuffer() = default;
```
Comment (roshandathathri) [inline libdist/include/galois/runtime/Serialize.h:186]: I'm surprised that this isn't needed?

File: libdist/include/galois/runtime/Serialize.h [hunk 6]
```diff
@@ -174,27 +148,19 @@ class DeSerializeBuffer {
...
-
-  //! Moves DeserializedBuffer into SerializeBuffer
-  SerializeBuffer& operator=(DeSerializeBuffer&& rhs) {
-    bufdata = std::move(rhs.getVec());
+  using size_type = vTy::size_type;
+
+  //! Moves DeserializedBuffer into SendBuffer
+  SendBuffer& operator=(RecvBuffer&& rhs) {
+    auto buf = std::move(rhs);
+    bufdata = std::move(buf.getVec());
```
Comment (roshandathathri) [inline libdist/include/galois/runtime/Serialize.h:163]: What's the difference b/w this and `rhs.getVec()`?

**Review by roshandathathri — APPROVED** [review]
This is great!

Some minor nits below.

File: libdist/include/galois/runtime/Network.h (hunk 3 above)
Comment (ddn0) [reply libdist/include/galois/runtime/Network.h:46]: It's a little hard to refer to if it is in the NetworkInterface class:

```
galois::runtime::NetworkInterface::RecvResult
// or should I use the concrete subclass?
galois::runtime::NetworkInterfaceBuffered::RecvResult
```
vs
```
galois::runtime::RecvResult
```
which makes it clear that this is independent of the implementation class.

This is not big deal to me, but that was my reasoning.

File: libdist/include/galois/runtime/Serialize.h (hunk 4 above)
Comment (ddn0) [reply libdist/include/galois/runtime/Serialize.h:60]: cool

**Review by gurbinder533 — APPROVED** [review]
LGTM.

File: libdist/include/galois/runtime/Network.h (hunk 3 above)
Comment (roshandathathri) [reply libdist/include/galois/runtime/Network.h:46]: Ok, makes sense. SendBuffer and RecvBuffer are in galois::runtime as well.

File: libdist/include/galois/runtime/Serialize.h (hunk 4 above)
Comment (roshandathathri) [reply libdist/include/galois/runtime/Serialize.h:60]: Cool

File: libdist/include/galois/runtime/Serialize.h (hunk 5 above)
Comment (roshandathathri) [reply libdist/include/galois/runtime/Serialize.h:186]: Thanks for the clarification.

File: libdist/include/galois/runtime/Serialize.h (hunk 6 above)
Comment (roshandathathri) [reply libdist/include/galois/runtime/Serialize.h:163]: Still confusing to me. `rhs` might be destructed earlier but then `buf` will only be destructed when the function returns, right?

Comment (ddn0) [reply libdist/include/galois/runtime/Serialize.h:163]:
```cpp
// Version 1
SendBuffer& operator=(RecvBuffer&& a) {
  bufdata = std::move(a.getVec());
  ...
}

// Version 2
SendBuffer& operator=(RecvBuffer&& a) {
  auto buf = std::move(a);
  bufdata = std::move(buf.getVec());
  ...
}

void Fn() {
  RecvBuffer rbuf;
  SendBuffer b = std::move(rbuf);
  .....
}
```

In version 1, if RecvBuffer had any other fields besides `getVec` they would not be released until `Fn` returns. In version 2, those fields would be released when `operator=(RecvBuffer&&)` returns.

Comment (roshandathathri) [reply libdist/include/galois/runtime/Serialize.h:163]: Thanks for the clarification. Didn't know this. I thought that the move object passed as a function parameter would be destructed at the end of that function, which in this case is `operator=`.

Comment (ddn0) [reply libdist/include/galois/runtime/Serialize.h:163]: I can't claim to be an expert in move semantics, but the model that helps me is that `std::move` is just a cast to an rvalue, and an rvalue is just a type that helps dispatch to the operations that actually implement move behavior. It isn't until you use `operator=(T&&)` or similar that the magic happens. In fact, it is probably a bug to take an rvalue as a parameter and never `std::move` it or use a sink function like `operator=(T&&)`.

Comment (roshandathathri) [reply libdist/include/galois/runtime/Serialize.h:163]: Thanks!

File: libdist/include/galois/runtime/Network.h (hunk 2 above)
Comment (ddn0) [reply libdist/include/galois/runtime/Network.h]: Turns out it wasn't too much code, so I added this change as another commit at the end of this one.

Comment (roshandathathri) [reply libdist/include/galois/runtime/Network.h]: Awesome!

**Pull request merged**